sdk/typescript/introspector/scanner/utils.ts

```diff
@@ -164,8 +164,9 @@ type OptionalValue = {
 /**
  * Return true if the parameter is optional.
  *
- * This includes both optional value defines with `?` and value that
- * have a default value.
+ * This only includes optional value defines with `?` or `<string>|null`.
+ * If a value has a default but isn't defined with `?`, it's not considered
+ * optional but nullable in the context of Dagger.
```

Comment:
In TS, `<string> | null` on it's own (without a default) isn't optional. And you can’t have a default and mark with `?` at the same time.

```suggestion
 * This only means optionals marked with `?`, or "nullable" types defined
 * with `| null`, to match the API's meaning of "optional".
```

API optionals in TypeScript should be:
- `foo?: <type>` (optional)
- `foo?: <type> | null` (optional)
- `foo: <type> | null = <default>` (nullable with default)

## Non-optional nullable without default

What about `foo: string | null` (without default)?  Because that makes it optional in the API but required in TS. To allow omitting it, the SDK needs to set the input to `null` before invoking the function when it’s not set. However, the proper way to declare this is with `foo?: string` instead, but users may do the former anyway.

I’ve done this in Python (setting to `None` on `foo: str | None` if not set):

https://github.com/dagger/dagger/blob/2ff9d5623b0643e84268f475e05b15e1d9fb9e6b/sdk/python/src/dagger/mod/_resolver.py#L324-L334

Notice that `inputs.get(param.name)` returns `None` if key `param.name` doesn’t exist in dictionary `inputs`.

## Returning `OptionalValue`

I also think it’s strange that an `isOptional` named function returns a non-boolean like `OptionalValue`. Especially since the meaning of “optional” has changed here. You’re basically checking if a type is nullable, but returning the default if it exists.

In Python I’ve changed the internal functions so that it’s easier to understand for those reading the code, splitting the concepts of "nullable” and “has default” from “optional”:

https://github.com/dagger/dagger/blob/79a10c6181de271979545e4c387023af407d8016/sdk/python/src/dagger/mod/_arguments.py#L24-L25

Notice how `is_nullable` is used differently from `is_optional`, which matches the API's meaning:

https://github.com/dagger/dagger/blob/79a10c6181de271979545e4c387023af407d8016/sdk/python/src/dagger/mod/_converter.py#L88-L89

## Confusing, I know

It is confusing, but the above `with_optional` in the API actually means "nullable" in GraphQL, while `is_optional` in GraphQL means "nullable or has default".

It's also confusing that we're making a distinction between the meaning in Dagger vs the meaning in GraphQL. In truth the `with_optional` call is an intermediary representation that the Dagger API uses to produce an extension to the GraphQL schema. So while the `with_optional` is Dagger specific, the term "optional" as in "nullable or has default" is both true in Dagger and in GraphQL because everything that's true in GraphQL must be true in the Dagger API since it uses GraphQL.

It's only this intermediary TypeDef representation that's confusing the terminology.

Reply:
> foo?: <type> (optional)
foo?: <type> | null (optional)
foo: <type> | null = <default> (nullable with default)

What about `foo: <type> | null`? 

Check https://github.com/dagger/dagger/pull/6766#issuecomment-1976861517

Reply:
> I also think it’s strange that an isOptional named function returns a non-boolean like OptionalValue. Especially since the meaning of “optional” has changed here. You’re basically checking if a type is nullable, but returning the default if it exists.

It was a way to optimize the check because otherwise I need to go through the exploration process 2 times.

Reply:
> What about `foo: <type> | null`?

It's right there, in _Non-optional nullable without default_. Haven't you read it?

Reply:
However I was under the impression that omitting that value would throw an error, when I've confirmed in https://github.com/dagger/dagger/pull/6766#discussion_r1511422324 that it executes. So no special handling necessary in the SDK but it's still not what a user should use when the intent is to represent an optional value.